### DISCUSSION TITLE
Google Cloud (GCP) Classic vs Google Cloud Native

### DISCUSSION THREAD

**Discussion by 23ewrdtf** [pulumi/pulumi#12470] [General]
### What happened?

 - What's the difference between: [Google Cloud (GCP) Classic](https://www.pulumi.com/registry/packages/gcp/) and [Google Cloud Native](https://www.pulumi.com/registry/packages/google-native/) ? 
 - Are both will be always fully supported?
 - Will one of them be dropped at some point in the future?
 - Which one should I use to be safe from deprecations in the future?
 - Are both support GCP fully and in the same way? From what I tried this is not the case.

The note says: 
```
Google Cloud Native is in public preview. Breaking changes may be introduced in minor version releases.
[Google Cloud Classic](https://www.pulumi.com/registry/packages/gcp) remains fully supported and recommended for production use.
```

Does it mean the Google Cloud Classic will be dropped at some point and we will have to migrate to Google Cloud Native?

### Expected Behavior

GCP provider should be one or if both the differences should be fully documented.Darwin

### Steps to reproduce

Docs: [Google Cloud (GCP) Classic](https://www.pulumi.com/registry/packages/gcp/) and [Google Cloud Native](https://www.pulumi.com/registry/packages/google-native/)

### Output of `pulumi about`

✗  pulumi about
CLI
Version      3.58.0
Go Version   go1.20.1
Go Compiler  gc

Host
OS       darwin
Version  12.6.2
Arch     arm64

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by danielrbradley** [comment]
Hi @23ewrdtf here's where we are right now:

As you mention, google-native is currently in preview. This means there may be breaking changes and it's expected to be less stable than other providers. It also doesn't have a regular, guaranteed release cycle or bug SLA.

The pulumi-gcp provder is what we describe as a "bridged" provider because it shares implementation with Google's own terraform provider. Although it's described as "classic" in the registry - this is just indicating that it's bridged rather than native. This is a fully supported provider which we indend to maintain long-term.

For choosing between the providers, I would personally recommend using the "classic" provider for most day-to-day work as it's much more stable and widely tested. There might be certain features that the preview native provider recieves quicker than the classic provider - at which point you've got the option to mix and match the providers where needed to gain access to the best of both providers.

**Comment by 23ewrdtf** [comment]
Thank you @danielrbradley 

I guess my main concern is the long term support and not need to migrate all the infrastructure from classic to native at some point in the future.

**Reply by danielrbradley** [reply]
The classic gcp provider is on our highest level of provider support and there are no plans to require any form of migration in the foreseeable future.

**Reply by j14s** [reply]
Sorry to necro this topic...but then why does the Native provider exist if not to eventually suplant the Classic?

**Reply by 23ewrdtf** [reply]
I would like to know that too.

**Reply by danielrbradley** [reply]
The aim of Pulumi's native provider initiatives are to create more complete and consistent providers for popular clouds which can be delivered with minimal manual authoring of code. This has worked very well as an approach for Azure, Kubernetes and AWS where there are well-defined patterns of resources available as specifications. 

The preview of the Google Native provider has highlighted a number of challenges due to how GCP's API are designed. There's more use of imperative configuration methods (i.e. POST requests) to affect the values of specific properties of the resource. These are not always described in a formal method to indicate the property they affect. This leaves certain areas of the resources inaccessible.

Secondly, the GCP (classic) provider is maintained directly by Google themselves and has recently become something they're providing as a primary development experience, resulting in a provider of high quality and completeness.

Hope that provides a bit more of the context between the two providers.
